_posts/2018-07-10-Play Framework-in-Scala–Action-Composition.md

```diff
@@ -0,0 +1,256 @@
+
+<h2 id="chaining_actionbuilders"> Chaining ActionBuilders </h2>
+
+So, we have a LoggingActionBuilder that logs the requests, and we have a AuthenticationActionBuilder that authenticates the requests.
+
+What if I want to build Actions whose requests I want to both log an authenticate?
+
+<strong>1. Do I need to create a third ActionBuilder that does both?</strong>
+This has two downsides. Firstly it duplicates logic. If the way we authenticate a request changes, we would have to remember to change on both ActionBuilders.
+Secondly, we would violate the principle of single responsibility, by having the third ActionBuilder both logging and authenticating.
+<strong>2. Can I join ActionBuilders ad-hoc? </strong>
+This is what we want. It would mean we develop custom ActionBuilders, which can be used separately and independently, but that could also be combined together at will. 
+
+Jumping to the result, it should be something like the following sketch:
+
+<img class=" size-full wp-image-398 aligncenter" src="https://aerodatablog.files.wordpress.com/2018/06/chainingactionbuilders.png" alt="ChainingActionBuilders" width="2124" height="494" />
+
+From my contact with Actions, I get the impression Play's Actions were not planned for composition. The documentation is lacking and there is not really many developers discussing it on StackOverflow.
+Still, the problem is underpinned by method <code>andThen</code> of the ActionBuilder.
+
+```scala
+trait ActionBuilder[+R[_], B] extends ActionFunction[Request, R] {
+  ......
+  override def andThen[Q[_]](other: ActionFunction[R, Q]): ActionBuilder[Q, B] = new ActionBuilder[Q, B] {
+    ....
+    def invokeBlock[A](request: Request[A], block: Q[A] => Future[Result]) =
+      self.invokeBlock[A](request, other.invokeBlock[A](_, block))
+  }
+}
+```
+
+The first to notice is <code>andThen</code> returns another ActionBuilder. Meaning, no matter how many ActionBuilders you chain together, the result will also be an ActionBuilder.
+This is fortunate since we already know how to deal with ActionBuilders.
+
+The second thing is that the argument for the andThen is an ActionFunction.<small>[Probably indicating ActionBuilder chaining is not supposed to be a thing]</small>
+But what we are really interested in is chaining an ActionBuilder with other ActionBuilder, since our motivation is to be able to group two components together who can also work independently and alone (namely the Logging and Authentication ActionBuilders). 
+
+But of course ActionBuilder extends ActionFunction:
+
+```scala
+trait ActionBuilder[+L[_], B] extends ActionFunction[Request, L]
+```
+
+So here is the main point:
+Because on line 3 the left type parameter of the ActionFunction must be R[_], i.e. the type of the original ActionBuilder <strong>AND</strong> because an ActionBuilder is an ActionFuntion with the left type parameter to Request[_] then:
```

Comment:
Again weird formatting. Also I don't why but the line break on "main point: " isn't appearing


![image](https://user-images.githubusercontent.com/34140210/43082015-e6d4857a-8e8a-11e8-8f2d-9b48cfb74672.png)